x/stakeibc/keeper/lsm.go

```diff
@@ -1,12 +1,110 @@
+		return transfertypes.DenomTrace{}, errorsmod.Wrapf(types.ErrInvalidLSMToken, "denom trace not found for %s", denom)
+	}
+
+	return denomTrace, nil
+}
+
+// Parses the LSM token's IBC path (e.g. transfer/channel-0) and confirms the channel ID matches
+// the transfer channel of a supported host zone
+func (k Keeper) GetHostZoneFromLSMTokenPath(ctx sdk.Context, path string) (types.HostZone, error) {
+	// Validate path regex which confirms the token originated only one hop away (e.g. transfer/channel-0)
+	if !IsValidIBCPath(path) {
+		return types.HostZone{}, errorsmod.Wrapf(types.ErrInvalidLSMToken, "ibc path of LSM token (%s) cannot be more than 1 hop away", path)
+	}
+
+	// Remove the "transfer/" prefix
+	channelId := strings.ReplaceAll(path, transfertypes.PortID+"/", "")
+
+	// Confirm the channel is from one of Stride's supported host zones
+	for _, hostZone := range k.GetAllHostZone(ctx) {
+		if hostZone.TransferChannelId == channelId {
+			return hostZone, nil
+		}
+	}
+
+	return types.HostZone{}, errorsmod.Wrapf(types.ErrInvalidLSMToken,
+		"transfer channel-id from LSM token (%s) does not match any registered host zone", channelId)
+}
+
+// Parses the LSM token's denom (of the form {validatorAddress}/{recordId}) and confirms that the validator
+// is in the Stride validator set
+func (k Keeper) GetValidatorFromLSMTokenDenom(denom string, validators []*types.Validator) (types.Validator, error) {
+	// Denom is of the form {validatorAddress}/{recordId}
+	split := strings.Split(denom, "/")
+	if len(split) != 2 {
+		return types.Validator{}, errorsmod.Wrapf(types.ErrInvalidLSMToken,
+			"lsm token base denom is not of the format {val-address}/{record-id} (%s)", denom)
+	}
+	validatorAddress := split[0]
+
+	// Confirm validator is in Stride's validator set
+	for _, validator := range validators {
+		if validator.Address == validatorAddress {
+			return *validator, nil
+		}
+	}
+
+	return types.Validator{}, errorsmod.Wrapf(types.ErrInvalidLSMToken,
+		"validator (%s) is not registered in the Stride validator set", validatorAddress)
+}
+
+// Checks if we need to issue an ICQ to check if a validator was slashed
+// The query runs at periodic intervals defined by the ValidatorExchangeRateQueryInterval
+func (k Keeper) ShouldQueryValidatorExchangeRate(ctx sdk.Context, validator types.Validator, stakeAmount sdk.Int) bool {
+	params := k.GetParams(ctx)
+	queryInterval := sdk.NewIntFromUint64(params.ValidatorExchangeRateQueryInterval)
+
+	// If the query interval is disabled, do not submit a query
+	// This should not be possible with the current parameter validation
+	//  function which enforces that it's greater than 0
+	if queryInterval.IsZero() {
+		return false
+	}
+
+	oldProgress := validator.ProgressTowardsExchangeRateQuery
+	newProgress := validator.ProgressTowardsExchangeRateQuery.Add(stakeAmount)
+
+	// Submit query if the query interval checkpoint has been breached
+	return oldProgress.Quo(queryInterval).LT(newProgress.Quo(queryInterval))
```

Comment:
You said that `ProgressTowardsExchangeRateQuery` is reset in the callback. What does that mean? The current setup doesn't seem like it needs reseting (you just check `(old // internal) < (new % interval)`)

If it is reset in the callback and two ICQs whose callbacks would reset the ProgressTowardsExchangeRateQuery are in flight simultaneously, could the following happen:
```
> 1st query callback lands, sets Progress to 0; second query is executed on host but takes a while to be relayed back
> before 2nd lands, 90% of the progress is consumed by other liquid stakes
> now the 2nd query lands and resets progress to 0
> now 100% additional progress can be consumed before another query is issued, so we've sent 190% of the progress quota without another query
```

Reply:
great observation!! Originally it was spec'd to not reset, but was later updated as the resetting makes it a bit more transparent 

I think the scenario you laid out is possible, with a slight caveat, that the second user would have to LS while the first query is still in flight. But nonetheless the issue is still there.

I guess we have to either non-reset it or use the different ICQ option where we combine the callback data so that once the first query lands, it finishes the second user's LS.

Reply:
Can we avoid this issue by not reseting at all? 

The current setup doesn't seem like it needs reseting (you just check `(old // internal) < (new % interval)`). 

If adding reseting adds some transparency but introduces this issue and requires complexity (eg combining callbacks) to fix, I'd vote against reseting

Reply:
Yeah I just remembered that we want to avoid combining the ICQs so the users transactions remain independent. 
I'm onboard with not reseting. Do you think we should rename these variables in a way that's more clear?

```
Param: ValidatorExchangeRateQueryInterval
Validator Field: ProgressTowardsExchangeRateQuery
```

Reply:
Yeah `ValidatorExchangeRateQueryInterval` seems clear to me (wdyt?) and maybe `ProgressTowardsExchangeRateQuery` would be more clear as something like `CumulativeLSMTokenDeposits`